### PR TITLE
Multiple Parameter Lists

### PR DESCRIPTION
**Pull request by FlandiaYingman**

### CODE REVIEW

File: hkmc2/shared/src/main/scala/hkmc2/codegen/js/JSBuilder.scala [hunk 1]
```diff
@@ -99,11 +102,15 @@ class JSBuilder extends CodeBuilder:
...
           TODO("getters")
-        case TermDefn(syntax.Fun, sym, S(ps), bod) =>
-          val vars = ps.map(p => scope.allocateName(p.sym)).mkDocument(", ")
-          doc"function ${sym.nme}($vars) { #{  # ${body(bod)} #}  # }"
+        case TermDefn(syntax.Fun, sym, ParamList(_, ps) :: pss, bod) =>
+          val paramList = ps.map(p => scope.allocateName(p.sym)).mkDocument(", ")
+          val result = pss.foldRight(bod)({ 
+            case (ParamList(_, ps), block) => 
+              Return(Lam(ps, block), false)
+          })
```
Comment (LPTK) [inline hkmc2/shared/src/main/scala/hkmc2/codegen/js/JSBuilder.scala]:
```suggestion
          val result = pss.foldRight(bod):
            case (ParamList(_, ps), block) => 
              Return(Lam(ps, block), false)
```

File: hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala [hunk 2]
```diff
@@ -428,16 +428,14 @@ extends Importer:
...
-                case S(ts) => // Go through all parameter lists
-                  ts.foldLeft((Ls[Param](), newCtx1)):
-                    case ((ps, ctx), t) => params(t)(using ctx).mapFirst(ps ++ _)
-                  .mapFirst(some)
-                case N => (N, newCtx1)
+              val (pss, newCtx) = 
+                td.paramLists.foldLeft(Ls[ParamList](), newCtx1)({case ((pss, ctx), ps) => 
+                  val (qs, newCtx) = params(ps)(using ctx)
+                  (pss :+ ParamList(ParamListFlags.empty, qs), newCtx)
+                })
```
Comment (LPTK) [inline hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala]:
```suggestion
                td.paramLists.foldLeft(Ls[ParamList](), newCtx1):
                  case ((pss, ctx), ps) => 
                    val (qs, newCtx) = params(ps)(using ctx)
                    (pss :+ ParamList(ParamListFlags.empty, qs), newCtx)
```

File: hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala [hunk 3]
```diff
@@ -428,16 +428,14 @@ extends Importer:
...
               val b = rhs.map(term(_)(using newCtx))
               val r = FlowSymbol(s"‹result of ${sym}›", nextUid)
-              val tdf = TermDefinition(k, sym, ps,
-                td.signature.orElse(newSignatureTrees.get(id.name)).map(term), b, r)
+              val tdf = TermDefinition(k, sym, pss, td.signature.orElse(newSignatureTrees.get(id.name)).map(term), b, r)
```
Comment (LPTK) [inline hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala]: Why this change?

File: hkmc2/shared/src/main/scala/hkmc2/syntax/Tree.scala [hunk 4]
```diff
@@ -157,10 +157,9 @@ object PlainTup:
   def apply(fields: Tree*): Tree = Tup(fields.toList)
 
 object Apps:
-  def unapply(t: Tree): Opt[(Ident, Ls[Tup])] = t match
+  def unapply(t: Tree): Opt[(Tree, Ls[Tup])] = t match
```
Comment (LPTK) [inline hkmc2/shared/src/main/scala/hkmc2/syntax/Tree.scala]: If this always matches, you should change the return type to `S[(Tree, Ls[Tup])]`.

File: hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala (hunk 3 above)
Comment (FlandiaYingman) [reply hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala]: Forgot to clean this 🥲

File: hkmc2/shared/src/main/scala/hkmc2/syntax/Tree.scala [hunk 5]
```diff
@@ -205,31 +204,33 @@ private def getName(t: Tree, symNme: Opt[Tree]): (Opt[Tree], Diagnostic \/ Ident
...
-        (symName, R(id), N, N, S(sign))
-      // show(t: Tree): Str
-      case InfixApp(App(id: Ident, args), Keyword.`:`, ret) =>
-        (symName, R(id), S(args :: Nil), N, N)
-      // show[A](t: Tree[A]): Str
-      case InfixApp(App(App(id: Ident, typeParams: TyTup), args), Keyword.`:`, ret) =>
-        // val sign = S(InfixApp(typeParams, Keyword.`->`, InfixApp(args, Keyword.`->`, ret)))
-        (symName, R(id), S(args :: Nil), S(typeParams), N)
+  lazy val (symName, name, paramLists, typeParams, signature): (Opt[Tree], Diagnostic \/ Ident, Ls[Tup], Opt[Tree], Opt[Tree]) =
+    def rec(t: Tree, symName: Opt[Tree]): (Opt[Tree], Diagnostic \/ Ident, Ls[Tup], Opt[Tree], Opt[Tree]) = t match
```
Comment (LPTK) [inline hkmc2/shared/src/main/scala/hkmc2/syntax/Tree.scala]: I _try_ to keep things under 100 columns of width. Let's keep this on two lines!

(The line above should also be broken down, but that's another story.)

**Review by LPTK — APPROVED** [review]
LGTM, thanks!

Please address my comment.

Also, it looks like I missed one file when making https://github.com/hkust-taco/mlscript/commit/de0b19691dae86e1c6f34ac48199a67363a23c7b. To remove the inconsistency between the output of different machines, change:

```diff
-globalThis
+:slient
+globalThis
```

in `GlobalThis.mls`.

**Comment by FlandiaYingman** [comment]
Done

**Comment by LPTK** [comment]
What. There are no changes printed by `git diff-files -p` but it still returns exit code 1. No idea why.
